lib/concurrent/synchronization/java_pure_object.rb

```diff
@@ -32,6 +33,10 @@ def ns_signal
           JRuby.reference0(self).notify
           self
         end
+
+        def ensure_ivar_visibility!
+          # relying on undocumented behavior of JRuby, ivar access is volatile
```

Comment:
@thedarkone where did you find the volatile field read on ivar read? I've found only https://github.com/pitr-ch/jruby/blob/master/core/src/main/java/org/jruby/runtime/ivars/VariableAccessor.java#L90-L115 where I could not find one, which would indicate that there is no relationship between reads and writes implying that JRuby ivars are not volatile.

Let me try to summarize what I've found out:
- JRuby ivar read does not read any volatile field (like `varTableStamp`). https://github.com/pitr-ch/jruby/blob/master/core/src/main/java/org/jruby/runtime/ivars/VariableAccessor.java#L90-L115
- Var table creation and expansion is protected by read and write to volatile varTableStamp so that is volatile https://github.com/pitr-ch/jruby/blob/master/core/src/main/java/org/jruby/runtime/ivars/StampedVariableAccessor.java#L106-138
- ivar update goes through https://github.com/pitr-ch/jruby/blob/master/core/src/main/java/org/jruby/runtime/ivars/StampedVariableAccessor.java#L151-163 which has fence or `putObjectVolatile` but there is no fence or `getObjectVolatile` on ivar read path. (The fence and `putObjectVolatile` might be useless here.)

It seems that the documentation https://github.com/jruby/jruby/wiki/Concurrency-in-jruby#volatility corresponds with this. cc @headius, could you help with deciphering, I do not know the code-base well I can easily overlook something.